tests/integrations/server/server.rs

```diff
@@ -0,0 +1,68 @@
+use std::{sync::Arc, time::Duration};
+
+use grpcio::*;
+use grpcio_health::{proto::HealthCheckRequest, HealthClient, ServingStatus};
+use service::service_event::ServiceEvent;
+use test_pd::Server as MockServer;
+use tikv::config::TikvConfig;
+
+#[test]
+fn test_restart_grpc_service() {
+    fail::cfg("mock_force_uninitial_logger", "return").unwrap();
+    let check_heath_api = |max_retry, client: &HealthClient| {
+        let req = HealthCheckRequest {
+            service: "".to_string(),
+            ..Default::default()
+        };
+        for i in 0..max_retry {
+            let r = client.check(&req);
+            if r.is_err() {
+                assert!(i != max_retry - 1);
+                std::thread::sleep(Duration::from_millis(500));
+                continue;
+            }
+            let resp = r.unwrap();
+            assert_eq!(ServingStatus::Serving, resp.status);
+            break;
+        }
+    };
+    let (service_event_tx, service_event_rx) = tikv_util::mpsc::unbounded();
+    let sender = service_event_tx.clone();
+    let tikv_thread = std::thread::spawn(move || {
+        let dir = test_util::temp_dir("test_run_tikv_server", true);
+        let mut pd_server = MockServer::new(1);
+        let eps = pd_server.bind_addrs();
+        let mut config = TikvConfig::default();
+        config.log.level = slog::Level::Critical.into();
+        config.log.file.filename = "".to_string();
+        config.storage.data_dir = dir.path().to_str().unwrap().to_string();
+        config.pd.endpoints = vec![format!("{}:{}", eps[0].0, eps[0].1)];
+        server::server::run_tikv(config, service_event_tx, service_event_rx);
+
+        pd_server.stop();
+    });
+
+    let env = Arc::new(Environment::new(1));
+    let channel = ChannelBuilder::new(env).connect("127.0.0.1:20160");
+    let client: HealthClient = HealthClient::new(channel);
+    let req = HealthCheckRequest {
+        service: "".to_string(),
+        ..Default::default()
+    };
+    let max_retry = 30;
+    check_heath_api(max_retry, &client);
+    // PAUSE grpc service and validate.
+    sender.send(ServiceEvent::PauseGrpc).unwrap();
+    std::thread::sleep(Duration::from_secs(1));
+    let resp = client.check(&req);
```

Comment:
Better retry check, because pause grpc is executed asynchronously.

```rust
let start = Instant::now();
loop {
  if start.elapsed() > Duration::from_secs(5) { panic!(); }
  if client.check(&req).is_err() { break; }
}
```

Reply:
thx.